Add contributing guidelines (#85)
trunk-io/plugins: contributing.md

# Contribution

Thanks for contributing to trunk's default plugins! Read on to learn more.

- [Overview](#overview)
- [Release process](#releases)
- [Adding new linters](#linters)
- [Adding new actions](#actions)
- [Guidelines](#guidelines)
- [Docs](https://docs.trunk.io)

## Overview

We use this repository to provide our users with default linters and actions. trunk automatically
adds the following to users' trunk.yaml:

```yaml
plugins:
  sources:
    - id: trunk
      uri: https://github.com/trunk-io/plugins
      ref: <latest_release>
```

Plugins can also be loaded from local paths, as below:

```yaml
plugins:
  sources:
    - id: trunk
      local: </path/to/repo/root>
```

Run `trunk run toggle-local` to quickly toggle this setting in this repo.

Adding a plugin source lets users run `trunk check enable` or `trunk actions enable` with linters
and actions defined in that plugin. For more information, see our
[docs](https://docs.trunk.io/docs/plugins).

> If you want to request a new feature or need help, please see our
> [Features Page](https://features.trunk.io/) or our [Slack](https://slack.trunk.io/).

Reviewer: ```suggestion
If you have questions, please [stop by our community Slack]; and if you have a feature request, you can [file it here].
```

## Releases

> Plugins are released on a fairly regular cadence that is independent of PRs. Users will pick up

Reviewer: ```suggestion
`trunk-io/plugins` is released on a fairly regular cadence that is independent of PRs. Users will pick up
```

these configuration changes by running `trunk upgrade` to automatically update their plugin version
to the latest release.

```yaml
plugins:
  sources:
    - id: trunk
      uri: https://github.com/trunk-io/plugins

>       ref: v0.0.8 # will change to v0.0.9

Reviewer: ```suggestion
      ref: v0.0.8  # will change to v0.0.9 on next "trunk upgrade"
```

# Contribution

Thanks for contributing to Trunk's default plugins! Read on to learn more.

- [Overview](#overview)
- [Release process](#releases)
- [Adding new linters](#linters)
- [Adding new actions](#actions)
- [Guidelines](#guidelines)
- [Docs](https://docs.trunk.io)

## Overview

We use this repository to provide our users with default linters and actions. trunk automatically
adds the following to users' trunk.yaml:

```yaml
plugins:
  sources:
    - id: trunk
      uri: https://github.com/trunk-io/plugins
      ref: v0.0.8
```

Plugins can also be loaded from local paths, as below:

```yaml
plugins:
  sources:
    - id: trunk

>       local: /path/to/repo/root

Reviewer 2: ```suggestion
      local: </path/to/repo/root>
```

Perhaps this is better to make it clear it is a placeholder.

```

Run `trunk run toggle-local` to quickly toggle this setting.

Adding a plugin source lets users run `trunk check enable` or `trunk actions enable` with linters
and actions defined in that plugin. For more information, see our
[docs](https://docs.trunk.io/docs/plugins).

If you want to request a new feature or get help with a problem, please see our
[Features Page](https://features.trunk.io/) or our [Slack](https://slack.trunk.io/).

> ## Releases

Reviewer: This section still feels a bit detail-heavy and doesn't feel like it naturally answers a user's question.

Stepping back, think about the journey/flow for someone contributing to plugins:

1. uses a linter `magicwand`
2. discovers `trunk`
3. opens PR to trunk-io/plugins with `magicwand` linter definition
4. wants to use `magicwand` in their own repo

IMO, the purpose of this section is to explain to the user what the process for going from step 3 to step 4 is, so we'd want to explain:

* trunk-io/plugins has a release process that is independent of PRs; that is, we don't live at HEAD but rather pin the version of a plugin (explain that we do this for the same reason we pin linter versions)
  * also explain what `ref` field this is (e.g. "in the above example, we're pointing at v0.0.8 of plugins")
* once a plugins release containing the desired change is created, a user can pick those changes up using `trunk upgrade`

----

In a similar vein, there may be other user flows that you have in mind for someone reading this doc. Remember: always keep your audience in mind when writing.

(related/useful: https://nick.groenen.me/posts/the-4-types-of-technical-documentation/)

Plugin sources are tied to the version specified with the `ref` field. When users run
`trunk upgrade`, 3 fields in their trunk.yaml will be updated:

1. The CLI version will be updated to the latest trunk binary.
2. The plugin `ref` will be updated to the latest compatible release of this repo.
3. All enabled linters will be updated to their latest versions.

We aim to provide a fairly regular release cadence for this repository. Releases are tied to minimum
compatible versions of the trunk CLI, identified by the `required_trunk_version` field. Users will
not be able to load a plugin source until they have upgraded to a sufficient CLI version.

If you need a linter or action definition from this repo that hasn't been released yet, you can
change the `ref` field of a plugin source to be any `git` ref. However, we recommend sticking to our
official tagged releases unless absolutely necessary.

## Linters

> If you want to add a new linter integration with trunk, refer to our docs on
> [custom linters](https://docs.trunk.io/docs/check-custom-linters) and
> [custom parsers](https://docs.trunk.io/docs/custom-parsers), and work through the following:

Reviewer: ```suggestion
To add a new linter:
```

"If you want to" and "with trunk" are both redundant here: if a user is in these docs, they want to do so; and these docs are all in the context of trunk.

1. Create a directory inside `linters/` with the name of your new linter.
2. Inside this new directory, create the following structure. If you ran `trunk check` in this
   repository recently, some of these files should be automatically created for you:

   ```text
   linters/
   └─my-linter/
     │ plugin.yaml
     │ my_linter.test.ts
     │ readme.md (optional)
     │ my-config.json (optional)
     └─test_data/
        └─basic.in.py (with appropriate extension)
   ```

> 3. Modify plugin.yaml to define your new linter. Refer to other linters in this repository as
>    examples.

Reviewer: ```suggestion
3. Add your linter definition to `plugin.yaml` (consult the docs for [custom linters] and [custom parsers] to understand how it should be defined)
```

4. Making sure the plugin in `.trunk/trunk.yaml` is pointing to your local repository, run
   `trunk check enable <my-linter>` to enable your linter, and run `trunk check` to verify that the
   configuration is valid and that you get desired diagnostics. Running `trunk check --verbose` can
   help provide greater insights when debugging.
5. Add a few simple test cases to `my_linter.test.ts` to exercise your linter and generate

>    snapshots. Refer to [Testing Guidelines](tests/readme.md) for more information on writing and

Reviewer 2: Do we need to document `linter-test-helper` somewhere?

Author: Not sure on this point. I made a note about files should be created for you, but I think it's more of a nice-to-have and not really a robust tool we want to advertise too overtly? Open to thoughts.

Author: Post-discussion. No need to document now, until more robust

   running tests.
6. Run `trunk check` to lint your changes.

> 7. Commit your entire linter subdirectory and create a
>    [Pull Request](https://github.com/trunk-io/plugins/compare). We will review your changes as soon
>    as we can.

Reviewer: ```suggestion
7. Open a PR with your changes!
```

Anyone contributing should know all of this :)

1. Create a directory inside `actions/` with the name of your new action.
2. Inside this new directory, create the following structure:

   ```text
   actions/
   └─my-action/
     │ plugin.yaml
     └─readme.md
   ```

3. Modify plugin.yaml to define your new action. Refer to other actions in this repository as
   examples.
4. If necessary, add additional scripts in the same directory for your action to reference.
5. Making sure the plugin in .trunk/trunk.yaml is pointing to your local repository, run
   `trunk actions enable <my-action>` to enable your linter, and run `trunk run <my-action>` to
   verify that the configuration is valid and that you get desired results. Running
   `trunk actions history <my-action>` can help provide greater insights when debugging.
6. We have not yet defined a testing framework for plugin actions, but we are working to add one
   soon! Please briefly document your action in its readme.md and in your PR.
7. Run trunk check to lint your changes.
8. Commit your entire action subdirectory and create a Pull Request. We will review your changes as
   soon as we can.

## Guidelines

Please follow the guidelines below when contributing:

> - After defining a new linter or action, please add it to the repository's root
>   [readme.md](readme.md).

Reviewer: ```suggestion
- After defining a new linter or action, please add it to [`readme.md`](readme.md).
```

- [trunk](https://docs.trunk.io/docs/compatibility) supports running on most versions of Linux and
  macOS. If your linter only runs on certain OSs, refer to the example of

>   [stringslint](linters/stringslint/stringslint.test.ts).

Reviewer: This should be in the linters section.
